### PR TITLE
ELPP-1824 Add related articles

### PR DESCRIPTION
**Pull request by stephenwf**
Hi @thewilkybarkid 

This is the first pass of the functionality for the related articles.

### CODE REVIEW

File: src/Model/ArticleVersion.php [hunk 1]
```diff
@@ -3,10 +3,11 @@
 namespace eLife\ApiSdk\Model;
 
 use DateTimeImmutable;
+use eLife\ApiSdk\Collection\ArraySequence;
 use eLife\ApiSdk\Collection\Sequence;
 use GuzzleHttp\Promise\PromiseInterface;
 
-abstract class ArticleVersion implements Model
+abstract class ArticleVersion implements Model, Article
```
Comment (thewilkybarkid) [inline src/Model/ArticleVersion.php]: `Article` should extend `Model` instead.

File: src/Model/ArticleVersion.php [hunk 2]
```diff
@@ -57,7 +59,8 @@ public function __construct(
         PromiseInterface $issue,
         PromiseInterface $copyright,
         Sequence $authors,
-        Sequence $reviewers
+        Sequence $reviewers,
+        Sequence $relatedArticles = null
```
Comment (thewilkybarkid) [inline src/Model/ArticleVersion.php]: Shouldn't be nullable, instead the sequence would be empty.

File: src/Serializer/ArticleVersionNormalizer.php [hunk 3]
```diff
@@ -103,18 +108,25 @@ public static function articleClass(string $type, string $status = null)
...
+                ->then(function (Result $article) {
+                    return $article['relatedArticles'] ?? [];
+                }));
         } else {
             $complete = null;
 
             $data['abstract'] = promise_for($data['abstract'] ?? null);
 
-            $data['authors'] = new ArraySequence($data['authors']);
+            $data['authors'] = new ArraySequence($data['authors'] ?? []);
```
Comment (thewilkybarkid) [inline src/Serializer/ArticleVersionNormalizer.php]: Shouldn't be needed?

File: src/Serializer/ArticleVersionNormalizer.php [hunk 4]
```diff
@@ -50,6 +53,8 @@ public function __construct(ArticlesClient $articlesClient)
     public static function articleClass(string $type, string $status = null)
     {
         switch ($type) {
+            case 'external-article':
+                return ExternalArticle::class;
```
Comment (thewilkybarkid) [inline src/Serializer/ArticleVersionNormalizer.php]: Shouldn't be needed?

File: src/Serializer/ArticleVersionNormalizer.php [hunk 5]
```diff
@@ -260,6 +280,12 @@ private function denormalizeSnippet(array $article) : PromiseInterface
                 })->toArray();
             }
 
+            if ($object->getRelatedArticles()->notEmpty()) {
+                $data['relatedArticles'] = $object->getRelatedArticles()->map(function (Article $relatedArticle) use ($format, $context) {
+                    return $this->normalizer->normalize($relatedArticle, $format, $context);
```
Comment (thewilkybarkid) [inline src/Serializer/ArticleVersionNormalizer.php]: `$context['snippet']` should be set to `true`, though easier to use the helper like https://github.com/elifesciences/api-sdk-php/blob/master/src/Serializer/CollectionNormalizer.php#L170.

File: test/ApiTestCase.php [hunk 6]
```diff
@@ -768,6 +768,20 @@ private function createArticlePoAJson(string $id, bool $isSnippet = false, bool
                     ],
                 ],
             ],
+            'relatedArticles' => $isSnippet === false ? [
```
Comment (thewilkybarkid) [inline test/ApiTestCase.php]: Don't include this test here, it's removed below anyway.

File: src/Model/ArticleVersion.php [hunk 7]
```diff
@@ -223,4 +227,9 @@ public function getPdf()
     {
         return $this->reviewers;
     }
+
+    final public function getRelatedArticles(): Sequence
```
Comment (thewilkybarkid) [inline src/Model/ArticleVersion.php:233]: Add PHPDoc to make it `@return Sequence|Article[]` (like `getSubjects()`).

File: src/Serializer/ExternalArticleNormalizer.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            $data['articleTitle'],
+            $this->denormalizeJournal($data, $format, $context),
+            $data['authorLine'],
+            $data['uri']
+        );
+    }
+
+    public function supportsDenormalization($data, $type, $format = null)
+    {
+        return ExternalArticle::class === $type;
```
Comment (thewilkybarkid) [inline src/Serializer/ExternalArticleNormalizer.php]: Should also take `Article::class === $type && 'external' === ($data['type'] ?? 'unknown')`.

File: src/Serializer/ArticleVersionNormalizer.php [hunk 9]
```diff
@@ -154,6 +166,14 @@ public static function articleClass(string $type, string $status = null)
         $data['versionDate'] = !empty($data['versionDate']) ? DateTimeImmutable::createFromFormat(DATE_ATOM, $data['versionDate']) : null;
         $data['statusDate'] = !empty($data['statusDate']) ? DateTimeImmutable::createFromFormat(DATE_ATOM, $data['statusDate']) : null;
 
+        $data['relatedArticles'] = $data['relatedArticles']
+            ->map(Closure::bind(function (array $article) use ($format, $context) {
+                // Get type.
+                $type = $this->articleClass($article['type'], $article['status'] ?? null);
+                // Denorm.
+                return $this->denormalizer->denormalize($article, $type, $format, $context);
```
Comment (thewilkybarkid) [inline src/Serializer/ArticleVersionNormalizer.php]: This is overly complicated, `$type` can just be `Article::class` so there's no need to work it out (like https://github.com/elifesciences/api-sdk-php/blob/master/src/Serializer/CollectionNormalizer.php#L73-L81).

File: test/ApiTestCase.php (hunk 6 above)
Comment (stephenwf) [reply test/ApiTestCase.php]: It creates an infinite loop when calling `createArticlePoAJson()` and `createArticleVoRJson()` so I had to add that additional check

**Comment by stephenwf** [comment]
@giorgiosironi @thewilkybarkid 

Just to note with Symfony 3.2 out and the components we were using out of beta it looks like this has caused a lot of failures in the project. (~12% of tests). Tested with the master branch and a clean composer install. 

Can we remove the `composer.lock` from the `.gitignore` to avoid this in future?

**Comment by thewilkybarkid** [comment]
That would just cause the tests to pass, not actually fix the problem (which presumably would happen if something using it upgraded Symfony to 3.2).

@giorgiosironi, could you investigate what's happened next week?

**Comment by stephenwf** [comment]
That is true, but it that would get caught in a pull request and not break master branch too.

**Comment by thewilkybarkid** [comment]
Haven't got time to investigate further at the moment, but my local `master` branch has passed fine.

**Comment by thewilkybarkid** [comment]
@stephenwf Think this is complete now.

**Pull request merged**